[Bug] ng-packagr does not emit declaration files

## Type of Issue

```
[x] Bug Report
```

## Description

`ng-packagr` does not emit declaration file for all of the compilation targets.

This happens with `ng-packagr@14`. So this seems to be a regression.


#### How To Reproduce

Steps:

```sh
$ git clone git@github.com:single-spa/single-spa-angular
$ cd single-spa-angular
$ yarn --frozen-lockfile
$ yarn ng-packagr -p libs/single-spa-angular/ng-package.json
```

The output folder is `lib`. Go to `lib/src/public_api.d.ts` and see that `single-spa-angular.d.ts` does not exist. You'll see the following code:

```ts
// lib/src/public_api.d.ts
export { enableProdMode } from './prod-mode';
export { singleSpaAngular } from './single-spa-angular'; // Cannot find module './single-spa-angular' or its corresponding type declarations.
export { getSingleSpaExtraProviders } from './extra-providers';
```


#### Version Information

```
$ node_modules/.bin/ng-packagr --version
ng-packagr:            14.0.2
@angular/compiler:     14.0.2
rollup:                2.75.7
typescript:            4.7.4
```

--- a/src/lib/ts/cache-compiler-host.ts
+++ b/src/lib/ts/cache-compiler-host.ts
@@ -39,8 +39,9 @@ export function cacheCompilerHost(
     entryPoint.dependsOn(node);
   };
 
-  const { flatModuleFile, entryFile } = entryPoint.data.entryPoint;
+  const { flatModuleFile, destinationPath, entryFile } = entryPoint.data.entryPoint;
   const flatModuleFileDtsFilename = `${flatModuleFile}.d.ts`;
+  const flatModuleFileDtsPath = ensureUnixPath(path.join(destinationPath, flatModuleFileDtsFilename));
   const hasIndexEntryFile = path.basename(entryFile.toLowerCase()) === 'index.ts';
 
   return {
@@ -74,15 +75,18 @@ export function cacheCompilerHost(
       sourceFiles?: ReadonlyArray<ts.SourceFile>,
     ) => {
       if (fileName.endsWith('.d.ts')) {
-        if (hasIndexEntryFile && path.basename(fileName) === flatModuleFileDtsFilename) {
-          // In case the entry file is index.ts, we should not emit the `d.ts` which are a re-export of the `index.ts`.
-          // Because it will cause a conflict.
-          return;
+        if (fileName === flatModuleFileDtsPath) {
+          if (hasIndexEntryFile) {
+            // In case the entry file is index.ts, we should not emit the `d.ts` which are a re-export of the `index.ts`.
+            // Because it will cause a conflict.
+            return;
+          } else {
+            // Rename file to index.d.ts so that TypeScript can resolve types without
+            // them needing to be referenced in the package.json manifest.
+            fileName = fileName.replace(flatModuleFileDtsFilename, 'index.d.ts');
+          }
         }
 
-        // Rename file to index.d.ts so that TypeScript can resolve types without
-        // them needing to be referenced in the package.json manifest.
-        fileName = fileName.replace(flatModuleFileDtsFilename, 'index.d.ts');
         sourceFiles.forEach(source => {
           const cache = sourcesFileCache.getOrCreate(source.fileName);
           if (!cache.declarationFileName) {

